File: 29Items-Projects/7Algorithms/7-String-algorithms/src/main/java/com/stringalgorithms/algorithms/Manacher.java
package com.stringalgorithms.algorithms;

import java.util.Objects;

public final class Manacher {
    private Manacher() {
    }

    public static String longestPalindrome(String text) {
        Objects.requireNonNull(text, "text must not be null");
        if (text.isEmpty()) {
            return "";
        }

        char[] transformed = transform(text);
        int[] radii = new int[transformed.length];
        int center = 0;
        int right = 0;
        int bestCenter = 0;
        int bestRadius = 0;

        for (int index = 0; index < transformed.length; index++) {
            int mirror = 2 * center - index;
            if (index < right) {
                radii[index] = Math.min(right - index, radii[mirror]);
            }

            while (index - radii[index] - 1 >= 0
                    && index + radii[index] + 1 < transformed.length
                    && transformed[index - radii[index] - 1] == transformed[index + radii[index] + 1]) {
                radii[index]++;
            }

            if (index + radii[index] > right) {
                center = index;
                right = index + radii[index];
            }

            if (radii[index] > bestRadius) {
                bestRadius = radii[index];
                bestCenter = index;
            }
        }

        int start = (bestCenter - bestRadius) / 2;
        return text.substring(start, start + bestRadius);
    }

    private static char[] transform(String text) {
        char[] transformed = new char[text.length() * 2 + 1];
        for (int i = 0; i < transformed.length; i++) {
            transformed[i] = i % 2 == 0 ? '#' : text.charAt(i / 2);
        }
        return transformed;
    }
}

File: 29Items-Projects/7Algorithms/7-String-algorithms/python/string_algorithms/algorithms/manacher.py
def longest_palindrome(text: str) -> str:
    if text == "":
        return ""

    transformed = "#" + "#".join(text) + "#"
    radii = [0] * len(transformed)
    center = 0
    right = 0
    best_center = 0
    best_radius = 0

    for index in range(len(transformed)):
        mirror = 2 * center - index
        if index < right:
            radii[index] = min(right - index, radii[mirror])

        while (
            index - radii[index] - 1 >= 0
            and index + radii[index] + 1 < len(transformed)
            and transformed[index - radii[index] - 1]
            == transformed[index + radii[index] + 1]
        ):
            radii[index] += 1

        if index + radii[index] > right:
            center = index
            right = index + radii[index]

        if radii[index] > best_radius:
            best_radius = radii[index]
            best_center = index

    start = (best_center - best_radius) // 2
    return text[start : start + best_radius]
